nemo_skills/inference/generate_solutions.py

```diff
@@ -133,29 +134,50 @@ def generate_solutions(cfg: GenerateSolutionsConfig):
-                    output.update(original_data_point)
-                    if 'error_message' not in output:
-                        output['error_message'] = extract_error_message(output['generation'])
-
-                    if cfg.generation_key != "generation":
-                        output[cfg.generation_key] = output.pop("generation")
-
-                    fout.write(json.dumps(output) + "\n")
-                data_points = []
+            if len(data_points) == cfg.batch_size or idx == len(data) - 1:
+                if cfg.inference.use_batch_api:
+                    # Using batch API for generation
+                    # Accessing the model directly since CodeExecutionWrapper does not support batch_generate
+                    request_metadata = llm.model.batch_generate(
+                        prompts=[prompt.build_string(data_point) for data_point in data_points],
+                        tokens_to_generate=cfg.inference.tokens_to_generate,
+                    )
+                    with open(cfg.output_file, 'at', encoding='utf-8') as fout:
+                        for data_point in data_points:
+                            fout.write(json.dumps(data_point) + '\n')
+
+                    # saving the request id to be able to retrieve results when they are ready
+                    with open(cfg.output_file + '-batch-request-id', 'wt', encoding='utf-8') as batch_fout:
+                        batch_fout.write(json.dumps({'request_id': request_metadata.id}))
+                    LOG.info('Submitted batch evaluation request. Please wait for the results to be ready.')
+                    LOG.info('The current status and final results can be accessed through summarize_results.py')
+                    LOG.info('Request metadata: %s', str(request_metadata))
+
+                    # Clear data_points as we've submitted the batch
+                    data_points = []
+                else:
+                    # batch-computing the outputs
+                    outputs = llm.generate(
+                        prompts=[prompt.build_string(data_point) for data_point in data_points],
+                        stop_phrases=list(cfg.prompt.stop_phrases),
+                        **asdict(cfg.inference),
+                    )
+
+                    for output, original_data_point in zip(outputs, data_points):
+                        # to make it easier to follow up with evaluation and limit accidental errors, we are adding
+                        # all of the ground-truth data to the output file alongside the generated solutions
+                        output.update(original_data_point)
+                        if 'error_message' not in output:
+                            output['error_message'] = extract_error_message(output['generation'])
+
+                        if cfg.generation_key != "generation":
+                            output[cfg.generation_key] = output.pop("generation")
```

Comment:
We also need to use the right generation key for the batch api, can you please save it to the batch-request-id so that it can later be used to properly write data in the same way as non-batch api?

Reply:
Done